helix-core/src/main/java/org/apache/helix/spectator/RoutingTableProvider.java

```diff
@@ -91,6 +160,52 @@ public RoutingTableProvider() {
     return instanceSet;
   }
 
+  /**
+   * returns all instances for all resources from {resource} group that are in a specific {state}
+   *
+   * @param resource
+   * @param state
+   * @param groupingEnabled
+   *
+   * @return empty list if there is no instance in a given state
+   */
+  public Set<InstanceConfig> getInstances(String resource, String state, boolean groupingEnabled) {
+    if (!groupingEnabled) {
+      return getInstances(resource, state);
+    }
+
+    Set<InstanceConfig> instanceSet = null;
+    RoutingTable _routingTable = _routingTableRef.get();
+    ResourceGroupInfo resourceGroupInfo = _routingTable.getResourceGroup(resource);
+    if (resourceGroupInfo != null) {
+      instanceSet = resourceGroupInfo.getInstances(state);
+    }
+    if (instanceSet == null) {
+      instanceSet = Collections.emptySet();
+    }
+    return instanceSet;
+  }
+
+  public Set<InstanceConfig> getInstances(String resource, String state,
+      List<String> resourceTags) {
+    Set<InstanceConfig> instanceSet = null;
```

Comment:
Refactored, now in ResourceGroupInfo, a resourceInfo is defined, instead of all its methods get re-implemented.